File: front/go/deepx/graph_viz.go
package deepx

import (
	"fmt"
	"strings"
)

func (g *Graph) ToDOT() string {
	var builder strings.Builder
	builder.WriteString("digraph computational_graph {\n")
	builder.WriteString("  rankdir=TB;\n")
	builder.WriteString("  node [shape=record];\n")

	// 遍历所有节点
	for _, node := range g.nodes {
		builder.WriteString(fmt.Sprintf(`"%p" `, node))
		switch node.Ntype() {
		case NodeTensor:
			// 张量节点：显示形状和梯度信息
			builder.WriteString(fmt.Sprintf(`[label= "%s \n %v`, node.Name(), node.(*TensorNode).Tensor().Shape))
			// if node.(*TensorNode).Tensor().requiresGrad {
			// 	builder.WriteString(`\n require_grad`)
			// }
			builder.WriteString(`"`)
			builder.WriteString(",shape=box")
			builder.WriteString(",labeljust=l")
			builder.WriteString(",color=skyblue")
			builder.WriteString(",style=filled")
			builder.WriteString(",fillcolor=aliceblue")
			builder.WriteString(",fontname=\"Sans-Serif\"")

		case NodeOp:
			// 操作节点：突出显示操作类型
			opNode := node.(*OpNode)
			builder.WriteString(fmt.Sprintf(`[label="%s"`, opNode.Shortchar()))
			builder.WriteString(",shape=box")
			builder.WriteString(",style=filled")
			builder.WriteString(",fillcolor=lightgray")
			builder.WriteString(",color=darkslategray")
			builder.WriteString(",fontname=\"Courier Bold\"")

		case NodeConstArg:
			// 常量参数节点：显示参数值
			constNode := node.(*ConstArgNode)
			var valueStr string
			switch constNode.argType {
			case ArgTypeInt:
				valueStr = fmt.Sprintf("%d", constNode.Int())
			case ArgTypeFloat:
				valueStr = fmt.Sprintf("%.2f", constNode.Float()) + "f"
			case ArgTypeString:
				valueStr = constNode.String()
			}
			builder.WriteString(fmt.Sprintf(`[label="%s"`, valueStr))
			builder.WriteString(",shape=diamond")
			builder.WriteString(",style=filled")
			builder.WriteString(",fillcolor=lightyellow")
			builder.WriteString(",color=goldenrod")
			builder.WriteString(",fontname=\"Sans-Serif\"")
		}
		builder.WriteString("];\n")
	}

	// 添加边连接，为边也添加样式
	for _, node := range g.nodes {
		for _, input := range node.Inputs() {
			builder.WriteString(fmt.Sprintf(`  "%p" -> "%p"`, input, node))
			builder.WriteString(`[color=gray40,penwidth=1.2,arrowsize=0.8]`)
			builder.WriteString(";\n")
		}
	}

	builder.WriteString("}\n")
	return builder.String()
}

File: front/py/deepx/autograd/graph_viz.py
import graphviz
from .graph import Graph
from ._tensornode import TensorNode
from ._opnode import OpNode
from ._constargnode import ConstArgNode

def graph_method(f):
    """装饰器:将函数注册为Graph类的方法"""
    # 延迟到模块加载完成后再绑定方法
    from .graph import Graph
    setattr(Graph, f.__name__, f)
    return f

@graph_method
def to_dot(self):
        """生成DOT格式的计算图可视化""" 
        dot = graphviz.Digraph(comment='Computational Graph')
        dot.attr(rankdir='TB')  # 从上到下的布局
        
        # 设置全局节点属性
        dot.attr('node', shape='record')
        
        # 添加节点
        for node in self.nodes:
            attrs = {
                'fontname': 'Sans-Serif',
                'labeljust': 'l'
            }
            
            if isinstance(node, TensorNode):
                # 张量节点：显示形状和梯度信息
                label = f"{node.name}\n{node.tensor().shape if node.tensor() else ''}"
                attrs.update({
                    'shape': 'box',
                    'color': 'skyblue',
                    'style': 'filled',
                    'fillcolor': 'aliceblue'
                })
                
            elif isinstance(node, OpNode):
                # 操作节点：突出显示操作类型
                label = node.shortchar()
                attrs.update({
                    'shape': 'box',
                    'style': 'filled',
                    'fillcolor': 'lightgray',
                    'color': 'darkslategray',
                    'fontname': 'Courier Bold'
                })
                
            elif isinstance(node, ConstArgNode):
                # 常量参数节点：显示参数值
                if node.arg_type == 'int':
                    value = str(node.get_int())
                elif node.arg_type == 'float':
                    value = f"{node.get_float():.2f}f"
                else:  # string
                    value = node.get_string()
                label = value
                attrs.update({
                    'shape': 'diamond',
                    'style': 'filled',
                    'fillcolor': 'lightyellow',
                    'color': 'goldenrod'
                })
                
            # 添加节点
            dot.node(str(id(node)), label, **attrs)
        
        # 添加边连接
        for node in self.nodes:
            for input_name, input_node in node.inputs().items():
                dot.edge(
                    str(id(input_node)), 
                    str(id(node)),
                    color='gray40',
                    penwidth='1.2',
                    arrowsize='0.8'
                )
        
        return dot 